Improve grappler layout optimizer to support 5D tensors

kaixih opened the pull request:
This PR further improves the grappler layout optimizer to support 5D tensors. Specifically, this PR does:

1. Fix the issue of the Conv3D/Conv3DBackpropInput/Conv3DBackpropFilter Transposers that fail to apply ScopedDataFormatUpgrader.
2. Improve the unit tests for (1).
3. Add support for 5D tensors for BiasAdd, BiasAddGrad, Shape, ShapeN, Slice, AddN, Concat.
4. Add new tests for (3).

fyi. @nluehr

kaixih committed:
Layout optimizer supports many other ops for 5D tensors

andyly commented on tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:
```diff
@@ -1090,7 +1130,8 @@ std::vector<int> LayoutAgnosticOpTransposer::GetVariadic4DFaninPorts(
-    if (IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 4) &&
+    if ((IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 4) ||
+         IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 5)) &&
         ((IsAfterDstToSrcTransform(context, *regular_fanin_node) &&
```
Does this support a mixture of say 4D and 5D tensors? I assume after changing the layout, only 5D tensors will be detected due to IsAfterDstToSrcTransform? And can a test case be added for this (ShapeN and IdentityN taking in ranked 4 and ranked 5 tensors)?

kaixih replied:
Can ShapeN take in tensors with different dims? I tried to give shape_n two different tensors but got `tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 5 and 4 for '{{node Add_2}} = Add[T=DT_INT32](ShapeN, ShapeN:1)' with input shapes: [5], [4].`

andyly replied:
It looks like ShapeN requires same ranks for all arguments. I don't believe IdentityN has that restriction though.

kaixih replied:
Done. Please check the new commit.

andyly commented on tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:
```diff
@@ -767,16 +767,41 @@ Status AvgPoolGradTransposer::TransposeNode(TransposeContext* context,
+  VLOG(3) << "GenericLayoutOptimizer: transforming node '" << node->GetName()
+          << "' with op '" << node->GetOp() << "' from data format '"
+          << context->src_format << "' to '" << context->dst_format << "'";
+  TF_RETURN_IF_ERROR(UpdateNode(context, node));
```
Can you add some comments explaining why data formats are updated after rewriting the node? Same with BiasAddGradTransposer.

kaixih replied:
Done. Please check the new commit.

kaixih committed:
Support IdentityN

andyly commented:
Currently this is failing for some XLA GPU tests (`//tensorflow/python/keras/layers:convolutional_test_xla_gpu`, `//tensorflow/python/keras/layers:convolutional_transpose_test_xla_gpu`) due to `DataFormatVecPermute` tf2xla kernel not being updated for 5D data formats https://github.com/tensorflow/tensorflow/blob/ab55c62645c7968cb2ebda456cc3ca85c4a411e9/tensorflow/compiler/tf2xla/kernels/data_format_ops.cc#L88-L162. Can you also update that tf2xla kernel?

kaixih committed:
Support XLA DataFormatVecPermute

kaixih commented:
@andyly PTAL.

kaixih commented:
It seems here is a rollback. Do we know which tests are broken? @andyly

andyly commented:
> It seems here is a rollback. Do we know which tests are broken? @andyly

Currently I am investigating the test that is failing. It is an internal test and I don't have easy access to running it under a certain configuration.

kaixih commented:
Any update on this? @andyly

andyly commented:
> Any update on this? @andyly

Let me check with the owner of the test again. I'll also try bisecting/splitting this PR (we can also try to split and land parts of this PR).

kaixih commented:
Sounds good. Please keep me posted and let me know if I can help with the bisecting/splitting.

andyly commented:
I got access to a repro and the issue seems to be coming from BiasAdd(Transposer). Has that op been updated for 5D data format support? From some time back I remembered the op kernel (at least on CPU) hard coding the channel dimension to 1 for NCHW and the last dimension for NHWC, but it supported 3D, 4D, and 5D tensors.

kaixih commented:
The BiasAdd transposer is newly added since we cannot directly use the DefaultLayoutSensitiveOpTransposer which does dataformatUpgrade -> updateNode -> updateFaninEdges -> updateFanoutEdges. However, the execution path we need in the BiasAdd is like updateNode -> dataformatUpgrade -> updateFaninEdges -> updateFanoutEdge (I have added some comments to explain why in the code.)

For both DefaultLayoutSensitiveOpTransposer and BiasAdd, we restrict the optimization only to rank = 4 or 5. As for the BiasAdd op, I don't think we have changed anything about it.

kaixih commented:
Not sure if you can paste the related log here so that we can think about the potential problem.

andyly commented:
Actually the error is in the ReduceTransposer. It hasn't been completely updated for 5D tensor support and was triggering an error transitively from the BiasAdd 5D transposes. Let me correct the ReduceTransposer.

andyly commented:
For some context, https://github.com/tensorflow/tensorflow/blob/89e81228719825b36d6278276efcd6fc146574a3/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L1439-L1443 needs to be updated also.

kaixih commented:
Awesome. This makes more sense.

Thanks for fixing the part. Do I need to submit a new PR or you could help re-open this PR from your side.

andyly commented:
I'll submit a quick fix for the ReduceTransposer and then we can try to roll forward your PR.

andyly commented:
The fix is landed in https://github.com/tensorflow/tensorflow/commit/28c5e97db99302890451b3d607a5a3572ee6ca2b

kaixih commented:
> The fix is landed in [28c5e97](https://github.com/tensorflow/tensorflow/commit/28c5e97db99302890451b3d607a5a3572ee6ca2b)

Thanks for the fix. Hope this fix + this PR can make those tests pass.

kaixih commented:
Do the internal tests pass now? Can we roll forward this PR? @andyly Thanks.

andyly commented:
> Do the internal tests pass now? Can we roll forward this PR? @andyly Thanks.

I think we are good to go with rolling forward this PR. Can you create it? Also, when I was scanning the PR (for debugging the failed test internally), there may still need to be a few edits (specifically `IsBiasAdd` allows for BiasAdd and BiasAddV1 but BiasAdd is the op that supports different data formats, so it needs to be restricted further).

kaixih commented:
Sure. Created #44270 which includes the changes about restricting optimizer to BiasAdd node.
